### PR TITLE
Extend transaction structure new fields

### PR DESCRIPTION
**Pull request by miiu96**
## Reasoning behind the pull request
- Add information about the gas used and the fee of a transaction
  
## Proposed changes
- Extended the `Transaction` structure that is returned on the `/transaction` endpoint with 2 extra fields: `gasUsed` and `fee`
- Changed the version of `mx-chain-core-go` module.
- PR from `mx-chain-core-go` repository: https://github.com/multiversx/mx-chain-core-go/pull/161

## Testing procedure
- Execute some transactions and check the `gasUsed` and the `fee` field if are returned by the `transaction` endpoint of the node and proxy

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -14,7 +14,7 @@ require (
 	github.com/google/gops v0.3.18
 	github.com/gorilla/websocket v1.5.0
 	github.com/mitchellh/mapstructure v1.5.0
-	github.com/multiversx/mx-chain-core-go v1.1.31
+	github.com/multiversx/mx-chain-core-go v1.1.32-0.20230131083243-ebccba0c240f
```
Comment (bogdan-rosianu) [inline go.mod:17]: proper tag before merging

File: node/external/transactionAPI/gasUsedAndFeeProcessor.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	tx.Fee = fee.String()
+
+	if tx.IsRelayed {
+		tx.GasUsed = tx.GasLimit
+		tx.Fee = tx.InitiallyPaidFee
+	}
+
+	hasRefund := false
+	for _, scr := range tx.SmartContractResults {
+		if scr.IsRefund {
```
Comment (bogdan-rosianu) [inline node/external/transactionAPI/gasUsedAndFeeProcessor.go:34]: might refactor:
```
if !scr.IsRefund {
   continue
}
....
break
}

File: node/external/transactionAPI/gasUsedAndFeeProcessor.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"math/big"
+
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-core-go/data/transaction"
+)
+
+type gasUsedAndFeeProcessor struct {
+	txFeeCalculator feeComputer
```
Comment (iulianpascalau) [inline node/external/transactionAPI/gasUsedAndFeeProcessor.go:11]: Calculator & Computer ? Let's rename to even-out the naming

File: node/external/transactionAPI/gasUsedAndFeeProcessor.go (hunk 2 above)
Comment (iulianpascalau) [reply node/external/transactionAPI/gasUsedAndFeeProcessor.go:34]: Right, or even better: split in a new function

File: node/external/transactionAPI/gasUsedAndFeeProcessor.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+func (gfp *gasUsedAndFeeProcessor) prepareTxWithResultsBasedOnLogs(
+	tx *transaction.ApiTransactionResult,
+	hasRefund bool,
+) {
+	if tx.Logs == nil {
+		return
+	}
+
+	for _, event := range tx.Logs.Events {
+		if core.WriteLogIdentifier == event.Identifier && !hasRefund {
```
Comment (iulianpascalau) [inline node/external/transactionAPI/gasUsedAndFeeProcessor.go:56]: split in more functions?

File: node/external/timemachine/fee/feeComputer.go [hunk 5]
```diff
@@ -56,6 +56,39 @@ func NewFeeComputer(args ArgsNewFeeComputer) (*feeComputer, error) {
 	return computer, nil
 }
 
+// ComputeGasUsedAndFeeBasedOnRefundValue computes gas used and fee based on the refund value, at a given epoch
+func (computer *feeComputer) ComputeGasUsedAndFeeBasedOnRefundValue(tx *transaction.ApiTransactionResult, refundValue *big.Int) (uint64, *big.Int) {
```
Comment (iulianpascalau) [inline node/external/timemachine/fee/feeComputer.go:60]: unit tests for the new functions?

File: node/external/transactionAPI/gasUsedAndFeeProcessor.go (hunk 2 above)
Comment (miiu96) [reply node/external/transactionAPI/gasUsedAndFeeProcessor.go:34]: refactored.

File: node/external/transactionAPI/gasUsedAndFeeProcessor.go (hunk 3 above)
Comment (miiu96) [reply node/external/transactionAPI/gasUsedAndFeeProcessor.go:11]: changed

File: node/external/transactionAPI/gasUsedAndFeeProcessor.go (hunk 4 above)
Comment (miiu96) [reply node/external/transactionAPI/gasUsedAndFeeProcessor.go:56]: done.

File: node/external/timemachine/fee/feeComputer.go (hunk 5 above)
Comment (miiu96) [reply node/external/timemachine/fee/feeComputer.go:60]: added

File: node/external/timemachine/fee/feeComputer_test.go [hunk 6]
```diff
@@ -36,6 +37,139 @@ func TestNewFeeComputer(t *testing.T) {
...
+			GasPrice: 1000000000,
+		},
+	}
+
+	gasUsed := uint64(2_000_00)
+	fee := computer.ComputeTxFeeBasedOnGasUsed(tx, gasUsed)
+	require.Equal(t, "200000000000000", fee.String())
+
+	// epoch 200
+	tx = &transaction.ApiTransactionResult{
```
Comment (bogdan-rosianu) [inline node/external/timemachine/fee/feeComputer_test.go:116]: could have kept the same transaction and only update the epoch

File: node/external/timemachine/fee/feeComputer_test.go [hunk 7]
```diff
@@ -36,6 +37,139 @@ func TestNewFeeComputer(t *testing.T) {
...
+			GasLimit: 10_000_000,
+			GasPrice: 1000000000,
+		},
+	}
+
+	gasLimit := computer.ComputeGasLimit(tx)
+	require.Equal(t, uint64(63_500), gasLimit)
+
+	// epoch 200
+	tx = &transaction.ApiTransactionResult{
```
Comment (bogdan-rosianu) [inline node/external/timemachine/fee/feeComputer_test.go:159]: could have kept the same transaction and only update the epoch

File: node/external/timemachine/fee/feeComputer_test.go (hunk 6 above)
Comment (miiu96) [reply node/external/timemachine/fee/feeComputer_test.go]: fixed

File: node/external/timemachine/fee/feeComputer_test.go (hunk 7 above)
Comment (miiu96) [reply node/external/timemachine/fee/feeComputer_test.go]: fixed

File: node/external/transactionAPI/gasUsedAndFeeProcessor.go [hunk 8]
```diff
@@ -53,17 +52,25 @@ func (gfp *gasUsedAndFeeProcessor) prepareTxWithResultsBasedOnLogs(
...
-		if core.SignalErrorOperation == event.Identifier {
-			fee := gfp.txFeeCalculator.ComputeTxFeeBasedOnGasUsed(tx, tx.GasLimit)
-			tx.GasUsed = tx.GasLimit
-			tx.Fee = fee.String()
-		}
+func (gfp *gasUsedAndFeeProcessor) setGasUsedAndFeeBaseOnLogEvent(tx *transaction.ApiTransactionResult, hasRefund bool, event *transaction.Events) {
+	if core.WriteLogIdentifier == event.Identifier && !hasRefund {
+		gasUsed, fee := gfp.feeComputer.ComputeGasUsedAndFeeBasedOnRefundValue(tx, big.NewInt(0))
+		tx.GasUsed = gasUsed
+		tx.Fee = fee.String()
```
Comment (iulianpascalau) [inline node/external/transactionAPI/gasUsedAndFeeProcessor.go:63]: missing return here?

**Review by gabi-vuls — APPROVED** [review]
System test passed

File: go.mod [hunk 9]
```diff
@@ -14,7 +14,7 @@ require (
 	github.com/google/gops v0.3.18
 	github.com/gorilla/websocket v1.5.0
 	github.com/mitchellh/mapstructure v1.5.0
-	github.com/multiversx/mx-chain-core-go v1.1.32-0.20230131083243-ebccba0c240f
+	github.com/multiversx/mx-chain-core-go v1.1.32
```
Comment (iulianpascalau) [inline go.mod:17]: :+1:

**Review by iulianpascalau — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]

**Pull request merged**